pkg/sentry/kernel/fd_table.go

```diff
@@ -219,18 +222,26 @@ func (f *FDTable) NewFDs(ctx context.Context, fd int32, files []*fs.File, flags
 			end = int32(lim.Cur)
 		}
 		if fd >= end {
+			// FD exceeds limit, no available FDs until old FDs get released.
+			f.next = end
 			return nil, syscall.EMFILE
 		}
 	}
 
 	f.mu.Lock()
 	defer f.mu.Unlock()
 
+	// From f.next to find available fd.
+	if fd < f.next {
+		fd = f.next
+	}
+
 	// Install all entries.
 	for i := fd; i < end && len(fds) < len(files); i++ {
 		if d, _, _ := f.get(i); d == nil {
 			f.set(i, files[len(fds)], flags) // Set the descriptor.
 			fds = append(fds, i)             // Record the file descriptor.
+			f.next = i + 1                   // Update next search start position.
```

Comment:
this looks wrong if f.next < fd.

I think we can do something like this:
```
                if d, _, _ := f.get(i); d == nil {
                        f.set(i, files[len(fds)], flags) // Set the descriptor.
                        fds = append(fds, i)             // Record the file descriptor.
-                       f.next = i + 1                   // Update next search start position.
                }
        }
 
@@ -250,13 +249,13 @@ func (f *FDTable) NewFDs(ctx context.Context, fd int32, files []*fs.File, flags
                for _, i := range fds {
                        f.set(i, nil, FDFlags{}) // Zap entry.
                }
-               if len(fds) != 0 {
-                       // fds is sorted, so simply assign the lowest fds[0] to f.next.
-                       f.next = fds[0]
-               }
                return nil, syscall.EMFILE
        }
 
+       if fds[0] == f.next {
+               f.next = fds[len(fds)-1] + 1
+       }
+
        return fds, nil
 }
```

Reply:
Yes, it's is my fault. I used to have some misunderstandings about FD.:joy:

Reply:
All others I agree with you, except I have one different idea about this:
```go
+       if fds[0] == f.next {
+               f.next = fds[len(fds)-1] + 1
+       }
```

`if fds[0] == f.next {` Maybe this `if condition` is a little strict, of couse it will work, but make `f.next` hard to update.
Is the following code looks better?
```go
+ if fd == f.next {
+         f.next = fds[len(fds)-1] + 1
+ }
```

Let me describe my thought:
1. if `fd` > `f.next`, we don't need to update `f.next`
2. else if `fd` <= `f.next` we need to update `f.next`

So, `fds[0] == f.next` is coverd in `fd == f.next`

```go
// From f.next to find available fd.
if fd < f.next {
    fd = f.next
}

// Install all entries.
for i := fd; i < end && len(fds) < len(files); i++ {
    if d, _, _ := f.get(i); d == nil {
        f.set(i, files[len(fds)], flags) // Set the descriptor.
        fds = append(fds, i)             // Record the file descriptor.
    }
}

// Failure? Unwind existing FDs.
if len(fds) < len(files) {
    for _, i := range fds {
        f.set(i, nil, FDFlags{}) // Zap entry.
    }
    return nil, syscall.EMFILE
}
if fd == f.next {
    f.next = fds[len(fds)-1] + 1
}
return fds, nil
```

Reply:
I'm ok with this variant too.